### PR TITLE
keras preprocessing layer random generator fix

### PR DESCRIPTION
**Pull request by yixingfu**
1. tracking state variable in stateful random generators;
2. apply the random generator fix on other Ranodm___ layers to RandomCrop;
3. add test coverage for training phase (previously only tested inference phase) for `image_preprocessing_distribution_test`

### CODE REVIEW

File: tensorflow/python/keras/layers/preprocessing/image_preprocessing_distribution_test.py [hunk 1]
```diff
@@ -42,8 +42,9 @@ class ImagePreprocessingDistributionTest(
   def test_distribution(self, distribution):
     # TODO(b/159738418): large image input causes OOM in ubuntu multi gpu.
     np_images = np.random.random((32, 32, 32, 3)).astype(np.float32)
-    image_dataset = dataset_ops.Dataset.from_tensor_slices(np_images).batch(
-        16, drop_remainder=True)
+    np_labels = np.random.random((32, 1)).astype(np.float32)
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/preprocessing/image_preprocessing_distribution_test.py:45]: using np.int64?

File: tensorflow/python/keras/layers/preprocessing/image_preprocessing.py [hunk 2]
```diff
@@ -1323,14 +1323,17 @@ def _create_variable(self, *args, **kwargs):
     # that it skips the distribution-strategy check. When we are inside a
     # distribution-strategy scope, variables.Variable will pick a proper
     # variable class (e.g. MirroredVariable).
-    return variables.Variable(*args, **kwargs)
+    v = variables.Variable(*args, **kwargs)
+    K.track_variable(v)
+    return v
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/preprocessing/image_preprocessing.py:1328]: Do you still need this?

File: tensorflow/python/keras/layers/preprocessing/image_preprocessing.py [hunk 3]
```diff
@@ -1323,14 +1323,17 @@ def _create_variable(self, *args, **kwargs):
...
 
 
 def make_generator(seed=None):
   if seed:
-    return _RandomGenerator.from_seed(seed)
+    rng = _RandomGenerator.from_seed(seed)
   else:
-    return _RandomGenerator.from_non_deterministic_state()
+    rng = _RandomGenerator.from_non_deterministic_state()
+  return rng
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/preprocessing/image_preprocessing.py:1336]: Do you still need this?

**Review by tanzhenyu — COMMENTED** [review]
Thank you for the fix!

File: tensorflow/python/keras/layers/preprocessing/image_preprocessing.py (hunk 3 above)
Comment (yixingfu) [reply tensorflow/python/keras/layers/preprocessing/image_preprocessing.py]: Reverted.

File: tensorflow/python/keras/layers/preprocessing/image_preprocessing_distribution_test.py (hunk 1 above)
Comment (yixingfu) [reply tensorflow/python/keras/layers/preprocessing/image_preprocessing_distribution_test.py:45]: Thanks for the suggestion. changed.

File: tensorflow/python/keras/layers/preprocessing/image_preprocessing.py (hunk 2 above)
Comment (tanzhenyu) [reply tensorflow/python/keras/layers/preprocessing/image_preprocessing.py:1328]: Is this comment addressed?

Comment (yixingfu) [reply tensorflow/python/keras/layers/preprocessing/image_preprocessing.py:1328]: I think this is needed, as the method override `_create_variable` in stateful_random_ops.Generator

**Comment by gbaned** [comment]
@yixingfu  Can you please resolve conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Review by tanzhenyu — DISMISSED** [review]

**Comment by gbaned** [comment]
@yixingfu  Can you please resolve conflicts? Thanks!

**Comment by yixingfu** [comment]
Merging a conflict dismissed the approval. Can you approve it again?  Thanks! @tanzhenyu

**Review by tanzhenyu — DISMISSED** [review]

**Comment by gbaned** [comment]
@yixingfu  Can you please address Ubuntu Sanity errors? Thanks!

**Comment by yixingfu** [comment]
With the modified the test,  `test_distribution_test_distribution_MirroredCPUAndGPU_mode_graph` and `test_distribution_test_distribution_CentralStorageCPUAndGPU_mode_graph` are failing.  One of them has `NotImplementedError`. These were skipped when I tried locally, and are only exposed with the CI build. What should we do? @tanzhenyu

**Review by tanzhenyu — DISMISSED** [review]

**Review by gbaned — DISMISSED** [review]

**Comment by tanzhenyu** [comment]
> With the modified the test, `test_distribution_test_distribution_MirroredCPUAndGPU_mode_graph` and `test_distribution_test_distribution_CentralStorageCPUAndGPU_mode_graph` are failing. One of them has `NotImplementedError`. These were skipped when I tried locally, and are only exposed with the CI build. What should we do? @tanzhenyu

Can you set tf.config.set_soft_device_placement(True) for these and re-submit?

**Comment by yixingfu** [comment]
> > With the modified the test, `test_distribution_test_distribution_MirroredCPUAndGPU_mode_graph` and `test_distribution_test_distribution_CentralStorageCPUAndGPU_mode_graph` are failing. One of them has `NotImplementedError`. These were skipped when I tried locally, and are only exposed with the CI build. What should we do? @tanzhenyu
> 
> Can you set tf.config.set_soft_device_placement(True) for these and re-submit?

Thanks for the suggestion. I have pushed the change. It's not convenient for me to run any test locally right now, can you do a kokoro:force-run or approve to see if it works?

**Review by tanzhenyu — DISMISSED** [review]

**Comment by yixingfu** [comment]
Oh sorry the importing was erroneous. Corrected.

**Review by tanzhenyu — APPROVED** [review]

**Comment by tanzhenyu** [comment]
Looks like something is failing?

**Comment by yixingfu** [comment]
> Looks like something is failing?

I didn't quite understand the failure log. To me it looks like the failures are not related to changes made here?

**Comment by gbaned** [comment]
@yixingfu Can you please resolve conflicts? Thanks!

**Comment by swghosh** [comment]
Any updates here?

**Comment by gbaned** [comment]
@yixingfu  Can you please resolve conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by swghosh** [comment]
Hey @tanzhenyu, should I look into this?

**Comment by gbaned** [comment]
@yixingfu Can you please resolve conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 29 days that this pull-request has stalled. Please create a new pull-request with the requested changes.
